src/prepared_statement.c

```diff
@@ -0,0 +1,388 @@
+				free(current->ps);
+			}
+			free(current);
+			break;
+		}
+	}
+
+	slog_noise(server, "prepared statement %u added to server cache, %d cached items", server_ps->ps->query_id, cached_query_count + 1);
+	HASH_ADD_INT(server->server_prepared_statements, query_id, server_ps);
+
+	return true;
+}
+
+bool handle_parse_command(PgSocket *client, PktHdr *pkt)
+{
+	PgSocket *server = client->link;
+	PgParsePacket pp;
+	PgServerPreparedStatement *server_ps = NULL;
+	PgClientPreparedStatement *client_ps = NULL;
+	PgPreparedStatement *ps;
+	PktBuf *buf;
+	bool found;
+	bool res;
+	struct OutstandingParsePacket *opp = NULL;
+
+	Assert(server);
+
+	if (!unmarshall_parse_packet(client, pkt, &pp))
+		return false;
+
+	/* update stats */
+	client->pool->stats.ps_client_parse_count++;
+
+	/* Lookup query in global hash */
+	ps = get_prepared_statement(&pp, &found);
+
+	/* Register statement on client */
+	client_ps = create_client_prepared_statement(pp.name, ps);
+	HASH_ADD_STR(client->client_prepared_statements, stmt_name, client_ps);
+
+	if (found) {
+		/* Such query was already prepared */
+		HASH_FIND_INT(server->server_prepared_statements, &ps->query_id, server_ps);
+		if (server_ps)
+		{
+			/* Statement was already prepared on this server, do not forward packet */
+			slog_debug(client, "handle_parse_command: mapping statement '%s' to '%s' (query '%s')",
+					   client_ps->stmt_name, ps->stmt_name, ps->query_with_params);
+
+			buf = create_parse_complete_packet();
+			res = pktbuf_send_immediate(buf, client);
+			pktbuf_free(buf);
+			return res;
+		}
+	}
+	/* Statement was not prepared on this server, sent modified P packet */
+	slog_debug(client, "handle_parse_command: creating mapping for statement '%s' to '%s' (query '%s')",
+			   client_ps->stmt_name, ps->stmt_name, ps->query_with_params);
+
+	buf = create_parse_packet(ps->stmt_name, &pp);
+
+	/* update stats */
+	client->pool->stats.ps_server_parse_count++;
+
+	/* Track Parse command sent to server */
+	opp = calloc(sizeof *opp, 1);
+	opp->ignore = false;
+	list_append(&server->server_outstanding_parse_packets, &opp->node);
+
+	/* Register statement on server */
+	server_ps = create_server_prepared_statement(ps);
+	if (!register_prepared_statement(server, server_ps))
+		return false;
+
+	return pktbuf_send_queued(buf, server);
+}
+
+bool handle_bind_command(PgSocket *client, PktHdr *pkt)
+{
+	PgSocket *server = client->link;
+	PgBindPacket bp;
+	PgParsePacket pp;
+	PgClientPreparedStatement *client_ps = NULL;
+	PgServerPreparedStatement *server_ps = NULL;
+	PgPreparedStatement* ps;
+	PktBuf *buf;
+	struct OutstandingParsePacket *opp = NULL;
+	int diff;
+	uint32_t len;
+
+	Assert(server);
+
+	if (!unmarshall_bind_packet(client, pkt, &bp))
+		return false;
+
+	/* update stats */
+	client->pool->stats.ps_bind_count++;
+
+	HASH_FIND_STR(client->client_prepared_statements, bp.name, client_ps);
+	if (!client_ps) {
+		disconnect_client(client, true, "prepared statement '%s' not found", bp.name);
```

Comment:
This should not disconnect, instead it should return an ErrorResponse.

Reply:
Done